src/presets.rs

```diff
@@ -90,6 +93,45 @@ pub fn generate_safe_prime_with_rng<T: Integer + RandomBits + RandomMod>(
     }
 }
 
+#[cfg(feature = "rayon")]
+/// Use [`rayon`] to parallelize the prime search.
+///
+/// Returns a random prime of size `bit_length` using the provided RNG.
+///
+/// Panics if `bit_length` is less than 2, or greater than the bit size of the target `Uint`.
+///
+/// Panics if the platform is unable to spawn threads.
+pub fn par_generate_prime_with_rng<T>(rng: &mut (impl CryptoRngCore + Send + Sync + Clone), bit_length: u32) -> T
+where
+    T: Integer + RandomBits + RandomMod,
+{
+    if bit_length < 2 {
+        panic!("`bit_length` must be 2 or greater.");
+    }
+    let bit_length = NonZeroU32::new(bit_length).expect("`bit_length` should be non-zero");
+
+    // TODO(dp): decide how to set the threadcount.
+    let threadcount = core::cmp::max(2, num_cpus::get() / 2);
+    let threadpool = rayon::ThreadPoolBuilder::new()
+        .num_threads(threadcount)
+        .build()
+        .expect("If the platform can spawn threads, then this call will work.");
+    let start = random_odd_integer::<T>(rng, bit_length).get();
+    let sieve = Sieve::new(start, bit_length, false);
+
+    threadpool.install(|| {
+        match sieve.par_bridge().find_any(|c| {
+            //TODO(dp): This clone feels dumb. OTOH it is fairly cheap to clone a rng.
+            let mut rng = rng.clone();
+            is_prime_with_rng(&mut rng, c)
+        }) {
+            // TODO(dp): This clone is very annoying.
+            Some(p) => p.clone(),
+            None => par_generate_prime_with_rng(rng, bit_length.get()),
```

Comment:
Do you hit this line in your tests? I wonder how does creating a thread pool within a thread pool work.

Reply:
It happens. Not often, but for 8192bit primes it happens regularly. If we decide to go ahead with this feature I'll make sure to investigate further how/when the threadpool drops.

Reply:
So this was a good catch: the threads are not dropped when recursing. I reshuffled things a bit so I can do an explicit `drop` on the threadpool and checked that it works.